src/crimson/os/seastore/async_cleaner.h

```diff
@@ -1403,6 +1409,52 @@ class AsyncCleaner : public SegmentProvider, public JournalTrimmer {
   journal_type_t get_journal_type() const {
     return ecb->get_journal()->get_type();
   }
+  uint64_t get_len_to_head(journal_seq_t tail_seq) const {
+    ceph_assert(get_journal_type() ==
+      journal_type_t::CIRCULARBOUNDED_JOURNAL);
+    if (tail_seq == JOURNAL_SEQ_NULL) {
+      return 0;
+    }
+    auto head_seq = journal_head;
+    auto &head = head_seq.offset.as_blk_paddr();
+    auto &tail = tail_seq.offset.as_blk_paddr();
+    ceph_assert(head_seq >= tail_seq);
+    auto gap = head.get_block_off() > tail.get_block_off() ?
+      head.get_block_off() - tail.get_block_off() :
+      tail.get_block_off() - head.get_block_off();
+    return gap;
+  }
+  journal_seq_t get_tail_target_behind_head(journal_seq_t head,
+    uint64_t len) const {
+    if (get_journal_type() == journal_type_t::SEGMENT_JOURNAL) {
+      journal_seq_t seq = head;
+      if (head.segment_seq >= (len / segments.get_segment_size())) {
+	seq.segment_seq -= (len / segments.get_segment_size());
+      } else {
+	seq.segment_seq = 0;
+	seq.offset = P_ADDR_MIN;
+      }
+      return seq;
+    } else {
+      ceph_assert(get_journal_type() ==
+	journal_type_t::CIRCULARBOUNDED_JOURNAL);
+      auto &blk_addr = head.offset.as_blk_paddr();
+      if (blk_addr.get_block_off() < len && head.segment_seq > 0) {
+	return journal_seq_t {
+	  head.segment_seq - 1,
+	  paddr_t::make_blk_paddr(
+	    head.offset.get_device_id(),
+	    len - blk_addr.get_block_off())};
```

Comment:
Hmm, doesn't look correct. Could you split this commit into at least 3 parts? 

* Calculating targets (I'll work on a PR for rebase); 
* Changes about writing journal header; 
* And enable trimming in UT.
* ... (seems no more independent topics)